Update README.md (#194)
math-comp/hierarchy-builder: README.md

# Hierarchy Builder

Hierarchy Builder (HB) provides high level commands to declare a hierarchy of algebraic structure
(or interfaces if you prefer the glossary of computer science) for the Coq system.

Given a structure one can develop its theory, and that theory becomes automatically applicable to
all the examples of the structure. One can also declare alternative interfaces, for convenience

> or backward compatibility, and provide glue code linking thse interfaces to the structures part of

Reviewer: ```suggestion
or backward compatibility, and provide glue code linking these interfaces to the structures part of
```

the hierarchy.

HB commands compile down to Coq modules, sections, records, coercions, canonical structure instances
and notations following the *packed classes* discipline which is at the core of the [Mathematical
Components](https://github.com/math-comp/math-comp) library. All that complexity is hidden behind
a few concepts and a few declarative Coq commands.

## Example

```coq
From HB Require Import structures.
From Coq Require Import ssreflect ZArith.

> HB.mixin Record is_AddComoid A := {

Reviewer: ```suggestion
HB.mixin Record IsAddComoid A := {
```

additive, commutative, monoid.

```coq

> HB.mixin Record is_AbelianGrp A of AddComoid A := {

Reviewer: ```suggestion
HB.mixin Record AddComoid_IsAbelianGrp A of AddComoid A := {
```

  opp : A -> A;
  addNr : forall x, opp x + x = 0;
}.

> HB.structure Definition AbelianGrp := { A of is_AbelianGrp A & is_AddComoid A }.

Reviewer: ```suggestion
HB.structure Definition AbelianGrp := { A of AddComoid_IsAbelianGrp A & IsAddComoid A }.
```

```coq
HB.instance Definition Z_CoMoid :=
  is_AddComoid.Build Z 0%Z Z.add Z.add_assoc Z.add_comm Z.add_0_l.
 
HB.instance Definition Z_AbGrp :=

>   is_AbelianGrp.Build Z Z.opp Z.add_opp_diag_l.

Reviewer: ```suggestion
  IsAbelianGrp.Build Z Z.opp Z.add_opp_diag_l.
```
